src/ReachSets/reach.jl

```diff
@@ -211,6 +214,130 @@ function reach(S::AbstractSystem,
+               assume_sparse=true,
+               assume_homogeneous=false,
+               numeric_type::Type=Float64,
+               lazy_X0=false,
+               kwargs...)::Vector{<:LazySet}
+               # unpack arguments
+               kwargs_dict = Dict(kwargs)
+
+
+
+               # list containing the arguments passed to any reachability function
+               args = []
+               waiting_list = []
+               # coefficients matrix
+               #TODO get start state. For now we assume that it is the first location
+               cur_loc_id = 1
+               cur_loc = HS.modes[init_loc_id]
+
+               #TODO push_initial_set
+               push!((cur_loc[1], X0), waiting_list)
+               i = 0
+               while (!isempty(waiting_list) and i < 15) #TODO add variable for max iteration number
+
+                   S = ContinuousSystem(cur_loc.A, X0, cur_loc.U)
+
+                   Rsets = reach(
+                        S,
+                        kwargs_dict[:N];
+                        algorithm=kwargs_dict[:algorithm],
+                        ε_init=kwargs_dict[:ε_init],
+                        set_type_init=kwargs_dict[:set_type_init],
+                        ε_iter=kwargs_dict[:ε_iter],
+                        set_type_iter=kwargs_dict[:set_type_iter],
+                        assume_sparse=kwargs_dict[:assume_sparse],
+                        assume_homogeneous=kwargs_dict[:assume_homogeneous],
+                        lazy_X0=kwargs_dict[:lazy_X0],
+                        blocks=kwargs_dict[:blocks],
+                        partition=kwargs_dict[:partition],
+                        block_types_init=kwargs_dict[:block_types_init],
+                        block_types_iter=kwargs_dict[:block_types_iter],
+                        template_directions_init=kwargs_dict[:template_directions_init],
+                        template_directions_iter=kwargs_dict[:template_directions_iter],
+                        lazy_inputs_interval=kwargs_dict[:lazy_inputs_interval],
+                        output_function=kwargs_dict[:output_function]
+                        )
+                       for (j, tj) in enumerate(transitions(HS)) #TODO optimize it or add method for outgoing transitions to SX
+
+                            if source(HS, tj) == cur_loc_id
+                                destination_loc = HS.modes[target(HS, tj)]
+                                source_invariant, target_invariant = cur_loc[2], destination_loc[2]
+                                #guard =
+                                ```
+                                get transition_info:
+                                Guard
+                                Destination_location
+                                # check intersection G_I^+_i^-
+                                if (intersection != empty)
+                                    push!(waiting_list, (inter_set, loc_id))
```

Comment:
`Rsets` is a flowpipe, i.e., a list of sets. It seems that you plan to take the intersection with the whole flowpipe. I am currently uncertain if that is good or not. The possible problem I see is that the intersection might be interrupted (e.g., intersection at steps 3-5 and at steps 7-8). This means that the intersection is either not convex or (potentially very) approximative. On the other hand, looking at each set in isolation is probably inefficient.
I propose to outsource this part about the intersection to a new function. This simplifies the `reach_d` function and allows us to experiment with different implementations later.

Reply:
So, as I know, the general approach to make an intersection is:
(i) compute continuous post operator
(ii) take an intersection of constructed flowpipes(set of images at time t_i) with invariants of I_currrent and I_destination and with the guard according to transition. The result is another flowpipe (set of images), which are subset of initial flowpipe. 
(iii) 
    (iii).1 - basic algorithm: Put the tuple of destination_location and every image to the waiting list(so you will have ``` n``` number of tuples at this stage, if you have ``` n``` number of images in your flowpipe_intersect). This approach has a problem, that number of elements in waiting list could growth exponentially. 
   (iii).2 - That`s why people then construct Convex Hull of all(or with some ratio) images inside flowpipe_intersect
   (iii).3 - There is more efficient way to combine them, which is called template hull, but I`m not sure how it works.

Reply:
@ (ii): You do not take the intersection witht the invariant of l_destination at this point. Here you only want to determine the states from which you can make a jump. If you have assignments/updates, the difference becomes clear (with no assignments/updates, you are right).

@ (iii): Good point, you are ahead of me here. For the first prototype, I would stick to a simple implementation like the first approach. But it is good to have these considerations in the design of the code.

Reply:
@ (ii) Yes, we need to make a reset prior to take intersection with destination location.